### DISCUSSION TITLE
possible to use one schema's virtual fields in another virtual field (same schema)?

### DISCUSSION THREAD

**Discussion by FlorianML** [Automattic/mongoose#11080] [Q&A]
Hello, so wondering if this was possible or not, not having any luck. I'm running into an issue with one of the virtual fields on a schema. 

We have set 2 virtuals on the schema, the first one relies on a document from a different collection and this is retrieving the document successfully. The issue is that the 2nd virtual (clientTaxYearReportCount) on the schema, uses a field that is returned from the first virtual in order to correctly populate the field. Added a snippet below to maybe better communicate the issue.

```
/**
 * Add virtual fields
 */
schema.virtual('clientTaxDetail', {
  foreignField: 'id',
  localField: 'clientAccountId',
  ref: 'TaxPayer',
  justOne: true,
});

schema.virtual('clientTaxYearReportCount', {
  foreignField: 'taxAccountId',
  localField: 'clientAccountId',
  ref: 'report',
  match: (doc) => ({ taxYear: doc.clientTaxDetail.taxYear, status: 'done' }),
  count: true,
});
```

the virtual `clientTaxYearReportCount` uses the `clientTaxDetail.taxYear` field to use to match documents from a different collection. When running, the `match` operation returns `clientTaxDetail` as undefined. Is there anyway to have the virtual fields available? I've tried adding { toJSON: true} and { toObject: true } but didn't help.

Thank you!

**Comment by vkarpov15** [comment]
It depends on how you call `populate()`. Remember that, if you call `populate('clientTaxDetail clientTaxYearReportCount')`, Mongoose will populate the two virtuals in _parallel_. So `clientTaxDetail` might not be defined yet when the `match` function runs.

If you wait until `clientTaxDetail` is populated before populating `clientTaxYearReportCount` as shown below, you'll get the correct result:

```javascript
  const res = await Test.findOne().populate('clientTaxDetail');
  await res.populate('clientTaxYearReportCount');
```

Below is a complete example:

```javascript
'use strict';

const mongoose = require('mongoose');
const { Schema } = mongoose;

run().catch(err => console.log(err));

async function run() {
  await mongoose.connect('mongodb://localhost:27017/test');
  
  const schema = new Schema({
    clientAccountId: String
  });
  
  schema.virtual('clientTaxDetail', {
    foreignField: 'id',
    localField: 'clientAccountId',
    ref: 'TaxPayer',
    justOne: true,
  });
  
  schema.virtual('clientTaxYearReportCount', {
    foreignField: 'taxAccountId',
    localField: 'clientAccountId',
    ref: 'Report',
    match: (doc) => ({ taxYear: doc.clientTaxDetail.taxYear, status: 'done' }),
    count: true, 
  });
  
  const Test = mongoose.model('Test', schema);
  const TaxPayer = mongoose.model('TaxPayer', new Schema({
    id: String,
    taxYear: String,
    clientAccountId: String
  }));
  const Report = mongoose.model('Report', new Schema({ taxAccountId: String, status: String, taxYear: String }));

  await Report.create({ taxAccountId: 'testid1', status: 'done', taxYear: '2021' });
  await TaxPayer.create({ id: 'testid2', taxYear: '2021', clientAccountId: 'testid1' });
  await Test.create({ clientAccountId: 'testid2' });

  const res = await Test.findOne().populate('clientTaxDetail');
  await res.populate('clientTaxYearReportCount');
  console.log(res.toObject({ virtuals: true }));
}
```

Another alternative would be to make `clientTaxYearReportCount` a populated virtual on the `TaxPayer` document, and using deep populate.
